File: scripts/check_skill_command_parity.py
#!/usr/bin/env python3
"""Fail CI when the runtime skill copies disagree about a command's ENTRY POINT (CR071).

Each skill exists three times -- `.pi/skills/`, `.claude/skills/`, and
`plugins/mirror-mind/skills/` -- and every strangler flip has to rewrite the
invocations in all three. Nothing enforced that, and by 2026-09-09 eleven skills
reached the front door on Pi while still calling `uv run python -m memory`
directly on Claude Code and the published plugin. `.claude/skills/mm-journeys`
was still invoking Python for `journeys`, a command flipped in DS3.

Nothing was broken -- Python answered correctly -- but the burn-down's "answers
from TS" was true for one runtime and not the others, TS-only fixes never
reached those users, and DS10 deletes the module those copies invoke.

What is checked, and what deliberately is not:

  * CHECKED -- `.claude` and `plugins/mirror-mind` are byte-identical. They are
    today, for all 25 skills, and it is the cheapest strong invariant available.
  * CHECKED -- for every command a skill documents, all three copies reach it
    through the same entry point (front door or Python).
  * NOT CHECKED -- argument spellings. `.pi` documents `memories [args]` where
    `.claude` documents `memories [--type TYPE] [--layer LAYER] ...`. Forcing
    those identical would flatten documentation for no correctness gain.
  * NOT CHECKED -- the frontmatter name, the per-runtime Usage sections, or the
    natural-language examples, which are Portuguese in `.pi` and English in the
    others. Those differences are deliberate.

Usage:
    python scripts/check_skill_command_parity.py
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

PI_ROOT = REPO_ROOT / ".pi" / "skills"
CLAUDE_ROOT = REPO_ROOT / ".claude" / "skills"
PLUGIN_ROOT = REPO_ROOT / "plugins" / "mirror-mind" / "skills"

FRONT_DOOR = "front-door"
PYTHON = "python"

# The two invocation forms a skill may document. The captured group is the
# command surface that follows -- `journeys`, `soul load [slug]`, and so on.
FRONT_DOOR_RE = re.compile(r"ts/src/frontDoor/cli\.ts\s+(?P<rest>.+)$")
PYTHON_RE = re.compile(r"uv run python -m memory\s+(?P<rest>.+)$")


def command_name(rest: str) -> str:
    """The command a documented invocation reaches, ignoring its arguments.

    `soul load [slug]` and `soul load <slug> --session-id X` are the same
    command for this check's purpose: what matters is which engine answers
    `soul`, not how the copy spells its arguments.
    """
    for token in rest.split():
        if token.startswith(("-", "<", "[", "$", '"')):
            break
        return token
    return rest.strip()


def entry_points(path: Path) -> dict[str, set[str]]:
    """Map each documented command to the set of entry points it is reached by."""
    found: dict[str, set[str]] = {}
    for line in path.read_text(encoding="utf-8").splitlines():
        for pattern, entry in ((FRONT_DOOR_RE, FRONT_DOOR), (PYTHON_RE, PYTHON)):
            match = pattern.search(line)
            if match:
                found.setdefault(command_name(match.group("rest")), set()).add(entry)
                break
    return found


def check_byte_identity(skill: str, problems: list[str]) -> None:
    claude = CLAUDE_ROOT / skill / "SKILL.md"
    plugin = PLUGIN_ROOT / skill / "SKILL.md"
    if not claude.exists() or not plugin.exists():
        problems.append(f"{skill}: missing a .claude or plugins/mirror-mind copy")
        return
    if claude.read_bytes() != plugin.read_bytes():
        problems.append(
            f"{skill}: .claude and plugins/mirror-mind copies are not byte-identical"
        )


def check_entry_points(skill: str, problems: list[str]) -> None:
    copies = {
        ".pi": PI_ROOT / skill / "SKILL.md",
        ".claude": CLAUDE_ROOT / skill / "SKILL.md",
        "plugins/mirror-mind": PLUGIN_ROOT / skill / "SKILL.md",
    }
    present = {label: path for label, path in copies.items() if path.exists()}
    if len(present) < 2:
        return

    maps = {label: entry_points(path) for label, path in present.items()}
    commands = sorted({command for mapping in maps.values() for command in mapping})

    for command in commands:
        by_label = {label: mapping.get(command) for label, mapping in maps.items()}
        documented = {label: entries for label, entries in by_label.items() if entries}
        if len(documented) < 2:
            # A command only one copy documents is a prose difference, not a
            # flip that missed a copy. Out of scope by design.
            continue
        distinct = {frozenset(entries) for entries in documented.values()}
        if len(distinct) > 1:
            detail = ", ".join(
                f"{label}={'+'.join(sorted(entries))}" for label, entries in documented.items()
            )
            problems.append(f"{skill}: `{command}` disagrees on entry point -> {detail}")


def main() -> int:
    if not PI_ROOT.is_dir():
        print(f"skill parity check: {PI_ROOT} not found", file=sys.stderr)
        return 1

    problems: list[str] = []
    skills = sorted(path.name for path in PI_ROOT.iterdir() if path.is_dir())
    for skill in skills:
        check_byte_identity(skill, problems)
        check_entry_points(skill, problems)

    if problems:
        print("skill command parity: DRIFT DETECTED\n")
        for problem in problems:
            print(f"  {problem}")
        print(
            "\nRemediation: a flip must update the invocation in ALL THREE skill copies.\n"
            "Switch the lagging copies to the same entry point `.pi` uses; leave argument\n"
            "spellings and the Portuguese/English examples alone (CR071)."
        )
        return 1

    print(f"skill command parity: clean -- {len(skills)} skills agree on every entry point.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: ts/scripts/checkSkillCommandParity.ts
#!/usr/bin/env node
// Fail CI when a skill copy reaches Python, or when the three copies disagree
// about a command's entry point. (CV22.DS10.US2, decision D6)
//
// The Node successor to `scripts/check_skill_command_parity.py`. The port is
// not cosmetic: a guard whose job is to assert that nothing invokes the Python
// interpreter cannot itself require that interpreter to run. After TS5 there
// would be nothing left to execute the Python version, and the check that
// proves Python is gone would be the last thing keeping it.
//
// The ALLOWLIST MECHANISM IS DELETED, not emptied. CR072 introduced it so that
// each remaining Python invocation had to name the story that owed its port;
// by the end of CV22.DS10.US2 every entry was paid off. Keeping an empty
// dictionary would invite the next story to add a row to it, which is exactly
// the escape hatch DS10's Skill Invocation Gate exists to close. The assertion
// is now absolute: no skill copy, and no file in the packaged plugin, may
// invoke `uv run python -m memory`.
//
// What is checked, and what deliberately is not:
//
//   * CHECKED -- `.claude` and `plugins/mirror-mind` skill copies are
//     byte-identical.
//   * CHECKED -- for every command a skill documents, all three copies reach
//     it through the same entry point.
//   * CHECKED -- ABSENCE of `uv run python -m memory` in all three skill
//     trees AND anywhere in `plugins/mirror-mind/` (manifest, commands, hooks,
//     skills), which is the packaged-plugin half of the gate: an installed
//     user resolves that directory, not this repository's `.pi/`.
//   * NOT CHECKED -- argument spellings, frontmatter names, per-runtime Usage
//     sections, or the Portuguese/English examples. Those differences are
//     deliberate (CR071).
//
// Usage:  node ts/scripts/checkSkillCommandParity.ts

import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, relative } from "node:path";

const REPO_ROOT = new URL("../..", import.meta.url).pathname.replace(/\/$/, "");
const PI_ROOT = join(REPO_ROOT, ".pi", "skills");
const CLAUDE_ROOT = join(REPO_ROOT, ".claude", "skills");
const PLUGIN_DIR = join(REPO_ROOT, "plugins", "mirror-mind");
const PLUGIN_ROOT = join(PLUGIN_DIR, "skills");

const FRONT_DOOR_RE = /ts\/src\/frontDoor\/cli\.ts\s+(?<rest>.+)$/;
const PYTHON_RE = /uv run python -m memory\s*(?<rest>.*)$/;

type EntryPoint = "front-door" | "python";

function exists(path: string): boolean {
  try {
    statSync(path);
    return true;
  } catch {
    return false;
  }
}

function read(path: string): string {
  return readFileSync(path, "utf8");
}

/** The command a documented invocation reaches, ignoring its arguments. */
function commandName(rest: string): string {
  for (const token of rest.split(/\s+/)) {
    if (token.startsWith("-") || token.startsWith("<") || token.startsWith("[")) break;
    if (token.startsWith("$") || token.startsWith('"')) break;
    return token;
  }
  return rest.trim();
}

function entryPoints(path: string): Map<string, Set<EntryPoint>> {
  const found = new Map<string, Set<EntryPoint>>();
  for (const line of read(path).split("\n")) {
    for (const [pattern, entry] of [
      [FRONT_DOOR_RE, "front-door"],
      [PYTHON_RE, "python"],
    ] as const) {
      const match = pattern.exec(line);
      if (match?.groups) {
        const command = commandName(match.groups.rest ?? "");
        const set = found.get(command) ?? new Set<EntryPoint>();
        set.add(entry);
        found.set(command, set);
        break;
      }
    }
  }
  return found;
}

function copiesOf(skill: string): Record<string, string> {
  return {
    ".pi": join(PI_ROOT, skill, "SKILL.md"),
    ".claude": join(CLAUDE_ROOT, skill, "SKILL.md"),
    "plugins/mirror-mind": join(PLUGIN_ROOT, skill, "SKILL.md"),
  };
}

function checkByteIdentity(skill: string, problems: string[]): void {
  const claude = join(CLAUDE_ROOT, skill, "SKILL.md");
  const plugin = join(PLUGIN_ROOT, skill, "SKILL.md");
  if (!exists(claude) || !exists(plugin)) {
    problems.push(`${skill}: missing a .claude or plugins/mirror-mind copy`);
    return;
  }
  if (!readFileSync(claude).equals(readFileSync(plugin))) {
    problems.push(`${skill}: .claude and plugins/mirror-mind copies are not byte-identical`);
  }
}

function checkEntryPoints(skill: string, problems: string[]): void {
  const present = Object.entries(copiesOf(skill)).filter(([, path]) => exists(path));
  if (present.length < 2) return;

  const maps = present.map(([label, path]) => [label, entryPoints(path)] as const);
  const commands = new Set<string>();
  for (const [, map] of maps) for (const command of map.keys()) commands.add(command);

  for (const command of [...commands].sort()) {
    const documented = maps
      .map(([label, map]) => [label, map.get(command)] as const)
      .filter((pair): pair is readonly [string, Set<EntryPoint>] => pair[1] !== undefined);
    // A command only one copy documents is a prose difference, not a flip that
    // missed a copy. Out of scope by design.
    if (documented.length < 2) continue;

    const shapes = new Set(documented.map(([, set]) => [...set].sort().join("+")));
    if (shapes.size > 1) {
      const detail = documented.map(([label, set]) => `${label}=${[...set].sort().join("+")}`);
      problems.push(`${skill}: \`${command}\` disagrees on entry point -> ${detail.join(", ")}`);
    }
  }
}

/** Every file under a directory, recursively. */
function walk(dir: string): string[] {
  const out: string[] = [];
  for (const name of readdirSync(dir)) {
    const path = join(dir, name);
    if (statSync(path).isDirectory()) out.push(...walk(path));
    else out.push(path);
  }
  return out;
}

/**
 * The gate's own assertion: the Python entry point is ABSENT, not merely
 * agreed upon. Scans the three skill trees and the WHOLE packaged plugin --
 * manifest, commands, hooks, skills -- because an installed user resolves that
 * directory rather than this repository's `.pi/`.
 */
function checkPythonAbsent(problems: string[]): void {
  const roots = [PI_ROOT, CLAUDE_ROOT, PLUGIN_DIR].filter(exists);
  for (const root of roots) {
    for (const path of walk(root)) {
      if (/\.(png|jpg|jpeg|gif|zip|ico)$/i.test(path)) continue;
      let text: string;
      try {
        text = read(path);
      } catch {
        continue;
      }
      text.split("\n").forEach((line, index) => {
        if (PYTHON_RE.test(line)) {
          problems.push(
            `${relative(REPO_ROOT, path)}:${index + 1} invokes \`uv run python -m memory\` — ` +
              "route it through the front door (CV22.DS10 Skill Invocation Gate)",
          );
        }
      });
    }
  }
}

export function main(): number {
  if (!exists(PI_ROOT)) {
    process.stderr.write(`skill parity check: ${PI_ROOT} not found\n`);
    return 1;
  }
  const problems: string[] = [];
  const skills = readdirSync(PI_ROOT)
    .filter((name) => statSync(join(PI_ROOT, name)).isDirectory())
    .sort();

  for (const skill of skills) {
    checkByteIdentity(skill, problems);
    checkEntryPoints(skill, problems);
  }
  checkPythonAbsent(problems);

  if (problems.length > 0) {
    process.stdout.write("skill command parity: DRIFT DETECTED\n\n");
    for (const problem of problems) process.stdout.write(`  ${problem}\n`);
    process.stdout.write(
      "\nRemediation: a flip must update the invocation in ALL THREE skill copies.\n" +
        "Switch the lagging copies to the same entry point `.pi` uses; leave argument\n" +
        "spellings and the Portuguese/English examples alone (CR071). There is no\n" +
        "allowlist: CV22.DS10.US2 paid off the last entry and deleted the mechanism,\n" +
        "because a Python invocation cannot be correct once Python is gone.\n",
    );
    return 1;
  }

  process.stdout.write(
    `skill command parity: clean -- ${skills.length} skills agree on every entry point, ` +
      "and no copy invokes Python.\n",
  );
  return 0;
}

if (import.meta.filename === process.argv[1]) {
  process.exit(main());
}
